### PR TITLE
Don't check the unfinished expression before dot as the whole expression

### PR DESCRIPTION
**Pull request by auduchinok**
A test PR to check what breaks with this change.

Consider an unfinished expression with a dot at the end: `x.`. In many cases the dot would be ignored and only `x` expression would be checked. It has bad consequences where the type of `x` is unified with the type that is expected in the context, despite it's an inner expression in another unfinished one:
```fsharp
let f x =
    if true then
        x. // `x` is inferred to have type `unit`, which is not expected
```

That unexpected `unit` type breaks completion for cases where `x` should infer to have some other type, like `string` here:
```fsharp
let g (s: string) = ()

let f1 x =
    if true then
        // user wants to type `Length`,
        // but there's no completion, since `x` is mistakenly considered `unit`
        x. 

    // an error about `unit` not being compatible with `string`
    g x

// the same issues:
let f2 x =
    x. // trying to use `x` members from code completion
    g x
```

With this change both cases work as expected (i.e. `x` doesn't have any constraints in the first case and code completion suggests `string` members in the second one).

Another case is applying an argument:

```fsharp
let g (i: int) = ()
let h (s: string) = ()

let f x =
    g x. // `x` is inferred to `int` instead of `string`
    h x // an error
```

### CODE REVIEW

**Comment by auduchinok** [comment]
I've removed a few other related cases that prevented a good type checker recovery.

Before, right sight is ignored:

Screenshot 2023-02-09 at 10 20 27

After, right sight is type checked with the unknown type of the missing expression after dot:

Screenshot 2023-02-09 at 10 21 09

**Comment by auduchinok** [comment]
Please merge https://github.com/dotnet/fsharp/pull/14739 first, and I'll update the tests here afterwards.

**Comment by psfinaki** [comment]
@auduchinok there you go :)

**Comment by auduchinok** [comment]
@psfinaki Thanks!

I've updated parser tests for unfinished expressions (`A.`, `A.B.`) and added more tests for various incomplete record field cases (the parsing got broken there in the initial implementation, and only Code Completion tests caught it).

It's ready now.

File: src/Compiler/SyntaxTree/SyntaxTreeOps.fs [hunk 1]
```diff
@@ -463,14 +471,8 @@ let mkSynDot mDot m l (SynIdent (r, rTrivia)) =
...
-let mkSynDotMissing mDot m l =
-    match l with
-    | SynExpr.LongIdent (isOpt, SynLongIdent (lid, dots, trivia), None, _) ->
-        // REVIEW: MEMORY PERFORMANCE: This list operation is memory intensive (we create a lot of these list nodes)
-        SynExpr.LongIdent(isOpt, SynLongIdent(lid, dots @ [ mDot ], trivia), None, m)
-    | SynExpr.Ident id -> SynExpr.LongIdent(false, SynLongIdent([ id ], [ mDot ], [ None ]), None, m)
-    | SynExpr.DotGet (e, dm, SynLongIdent (lid, dots, trivia), _) -> SynExpr.DotGet(e, dm, SynLongIdent(lid, dots @ [ mDot ], trivia), m) // REVIEW: MEMORY PERFORMANCE: This is memory intensive (we create a lot of these list nodes)
-    | expr -> SynExpr.DiscardAfterMissingQualificationAfterDot(expr, m)
+let mkSynDotMissing (mDot: range) (m: range) (expr: SynExpr) =
+    SynExpr.DiscardAfterMissingQualificationAfterDot(expr, mDot, unionRanges mDot m)
```
Comment (auduchinok) [inline src/Compiler/SyntaxTree/SyntaxTreeOps.fs:475]: I guess I could just inline this function into the grammar rules now?

Comment (auduchinok) [reply src/Compiler/SyntaxTree/SyntaxTreeOps.fs:475]: Or maybe not. There're similar one line functions. I think these should be changed together in a separate PR if we decide to remove them.

**Review by T-Gro — APPROVED** [review]

**Pull request merged**